File: myqueue.py
class MyQueue:

    def __init__(self, head=None, tail=None):

        self.head = head
        self.tail = tail

    # Enqueue: add Node to back of the queue
    def enqueue(self, val):

        # Instantiate new node with value
        n = Node(val)

        # If queue empty add node
        if self.head is None:

            # n should be Node, having value and link

            self.head = n
            self.tail = n

            # enqueue does not need to return a value
            return None

        else:
            # link for old tail becomes new node
            self.tail.link = n

            # tail points to new tail
            self.tail = n

    # Dequeue: remove from front of the queue and return value

    def dequeue(self):

        # Queue cannot be empty
        assert self.head, "Queue is empty."

        # Temporarily store value for node at head
        _value = self.head.value

        # Reassign head property to node
        # current head's link points to
        self.head = self.head.link

        # If final item removed
        # set tail to None as well
        if self.head is None:
            self.tail = None

        # return value
        return _value

File: myqueue.js
class MyQueue {

    constructor(head=null, tail=null) {
      this.head = head;
      this.tail = tail;
      }

    // Enqueue: add node to tail
    enqueue(val) {

      // Instantiate new node with  value
      var n = new Node(val);

      // if queue is empty
      if (this.head === null) {
        this.head = n;
        this.tail = n;
      } else {
        // if queue has members
        // link for old tail becomes new node
        this.tail.link = n;

        // tail points to new tail
        this.tail = n;
      }

      // enqueue does not need to return a value
      return;

    }
    // dequeue: remove node from head
    dequeue() {

      // if queue empty, return (undefined)
      if (this.head === null) {
        return;
      } else {

        // Temporarily store value for node at head
        var tempValue = this.head.value;

        // Reassign head property to node
        // current head's link points to
        this.head = this.head.link;

        // If final node removed
        // set tail to null as well

        if (this.head === null) {
          this.tail = null;
        }

        // Return value
        return tempValue;
      }
    }
  }
